### PR TITLE
Hailem/29 get rid handle event pattern

### LINKED ISSUE #29: (1) [widget] Maybe we can get rid of the handleEvent pattern
Now that we have a better template system, we can try to get rid of the obscure `handleEvent` pattern.

It's supposed to make templating easier for users by removing all the explicit handlers from the template, so you can do this in the template:

```html
<button id="#mybutton">Click me</button>
```

And in your component have a method like:

```ts
{
  handleClick(e: Event) {
    // handle click here
  }
}
```

But when you have more complex templates with multiple click points you may want to have more fine-grained control.

So, like with templates, let devs to decide the method names and the behavior.

If users want to extend a template, they can, but they will have to take care of not breaking anything…

```html
<button @click=${this.confirmButtonClicked}">Confirm</button>
```

And if you want to use the `handleEvent` pattern, you still can use it by just implementing that method in the class.

### PR DESCRIPTION
**Pull request by hailem**
fixes #29

### CODE REVIEW

File: packages/input/src/input.ts [hunk 1]
```diff
@@ -28,25 +28,30 @@ export class Input extends Widget {
     return !this._input || this._input.value.trim().length === 0;
   }
 
+  public constructor() {
+    super();
+    this.handleInput = this.handleInput.bind(this);
+  }
+
```
Comment (carlosescri) [inline packages/input/src/input.ts]: I think this is not necessary if you use `@input=…` in the template.

File: packages/pager/src/index.ts [hunk 2]
```diff
@@ -81,6 +77,18 @@ export class Pager extends ReactiveWidget {
   public constructor() {
     super();
     this.visuallyhidden = true;
+    this.handleClick = this.handleClick.bind(this);
+    this.handleChange = this.handleChange.bind(this);
```
Comment (carlosescri) [inline packages/pager/src/index.ts]: If you add `@event=…` to template this probably won't be required.

File: packages/pager/src/index.ts [hunk 3]
```diff
@@ -81,6 +77,18 @@ export class Pager extends ReactiveWidget {
...
+  }
+
+  public connectedCallback(): void {
+    super.connectedCallback();
+    this.addEventListener('click', this.handleClick);
+  }
+
+  public disconnectedCallback(): void {
+    this.removeEventListener('click', this.handleClick);
+    super.disconnectedCallback();
```
Comment (carlosescri) [inline packages/pager/src/index.ts]: Although repetitive, you can get rid of this if you add `@click=…` to the four buttons in the template.

File: packages/pager/src/index.ts [hunk 4]
```diff
@@ -96,26 +104,26 @@ export class Pager extends ReactiveWidget {
...
-        this.nextPage();
-      } else if (role === 'last-page') {
-        this.page = this._lastPage;
-      }
-    } else if (event.type === 'change') {
-      this.page = parseInt(this.input.value, 10);
+  public handleClick(event: Event): void {
+    const button: HTMLButtonElement = event
+      .composedPath()
+      .find(el => el instanceof HTMLButtonElement) as HTMLButtonElement;
```
Comment (carlosescri) [inline packages/pager/src/index.ts]: If you add `@click=…` to the template we won't have to _search_ the button, the button should be the element pointed by `event.currentTarget`.

File: packages/results/src/card.ts [hunk 5]
```diff
@@ -1,6 +1,6 @@
-import { property, TemplateResult, query, CSSResultArray } from 'lit-element';
+import { property, TemplateResult, query, CSSResultArray, supportsAdoptingStyleSheets } from 'lit-element';
```
Comment (carlosescri) [inline packages/results/src/card.ts]: I think you imported `supportsAdoptingStyleSheets` by mistake…

File: packages/input/src/input.ts (hunk 1 above)
Comment (hailem) [reply packages/input/src/input.ts]: off course

**Pull request merged**